### PR TITLE
2.4.0.rc1

### LINKED ISSUE #214: Intermittent indexing issue with ElasticSearch 2.4.0.RC1 and Grails 3.3.2
I am new to ElasticSearch, upgrading a Grails 2.4.4 application that used the Searchable Plugin to Grails 3.3.2 using the ElasticSearch plugin. 

I have four searchable domains. Often, application startup will fail with the following error on one of my domains. In this case, it is failing on my User domain, saying it is not a searchable domain class. Sometimes, the application will start without error and I am able to search on all four domains as expected. 

Here's the error:

    2018-01-03 20:06:29.343 [main] ERROR org.springframework.boot.SpringApplication - Application 
    startup failed
    java.lang.IllegalArgumentException: Class class com.xxx.server.User_$$_javassist_39 is not a searchable 
    domain class.
    	at org.springframework.util.Assert.notNull(Assert.java:134)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.springsource.loaded.ri.ReflectiveInterceptor.jlrMethodInvoke(ReflectiveInterceptor.java:1427)
	at org.codehaus.groovy.reflection.CachedMethod.invoke(CachedMethod.java:93)
	at groovy.lang.MetaMethod.doMethodInvoke(MetaMethod.java:325)
	at org.codehaus.groovy.runtime.callsite.StaticMetaMethodSite$StaticMetaMethodSiteNoUnwrap.invoke(StaticMetaMethodSite.java:133)
	at org.codehaus.groovy.runtime.callsite.StaticMetaMethodSite.call(StaticMetaMethodSite.java:91)
	at org.codehaus.groovy.runtime.callsite.CallSiteArray.defaultCall(CallSiteArray.java:48)
	at org.codehaus.groovy.runtime.callsite.StaticMetaMethodSite.call(StaticMetaMethodSite.java:96)
	at org.codehaus.groovy.runtime.callsite.AbstractCallSite.call(AbstractCallSite.java:133)
	at grails.plugins.elasticsearch.index.IndexRequestQueue.indexEntityKeyFromInstance(IndexRequestQueue.groovy:99)
	at grails.plugins.elasticsearch.index.IndexRequestQueue$indexEntityKeyFromInstance$1.callCurrent(Unknown Source)
	at org.codehaus.groovy.runtime.callsite.CallSiteArray.defaultCallCurrent(CallSiteArray.java:52)
	at grails.plugins.elasticsearch.index.IndexRequestQueue$indexEntityKeyFromInstance$1.callCurrent(Unknown Source)
	at grails.plugins.elasticsearch.index.IndexRequestQueue.addIndexRequest(IndexRequestQueue.groovy:84)
	at grails.plugins.elasticsearch.index.IndexRequestQueue$addIndexRequest$0.callCurrent(Unknown Source)
	at org.codehaus.groovy.runtime.callsite.CallSiteArray.defaultCallCurrent(CallSiteArray.java:52)
	at grails.plugins.elasticsearch.index.IndexRequestQueue$addIndexRequest$0.callCurrent(Unknown Source)
	at grails.plugins.elasticsearch.index.IndexRequestQueue.addIndexRequest(IndexRequestQueue.groovy:79)
	at grails.plugins.elasticsearch.index.IndexRequestQueue$addIndexRequest.call(Unknown Source)
	at org.codehaus.groovy.runtime.callsite.CallSiteArray.defaultCall(CallSiteArray.java:48)
	at grails.plugins.elasticsearch.index.IndexRequestQueue$addIndexRequest.call(Unknown Source)
	at grails.plugins.elasticsearch.ElasticSearchService$_doBulkRequest_closure2$_closure10.doCall(ElasticSearchService.groovy:284)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.springsource.loaded.ri.ReflectiveInterceptor.jlrMethodInvoke(ReflectiveInterceptor.java:1427)
	at org.codehaus.groovy.reflection.CachedMethod.invoke(CachedMethod.java:93)
	at groovy.lang.MetaMethod.doMethodInvoke(MetaMethod.java:325)
	at org.codehaus.groovy.runtime.metaclass.ClosureMetaClass.invokeMethod(ClosureMetaClass.java:294)
	at groovy.lang.MetaClassImpl.invokeMethod(MetaClassImpl.java:1022)
	at groovy.lang.Closure.call(Closure.java:414)
	at groovy.lang.Closure.call(Closure.java:430)
	at org.codehaus.groovy.runtime.DefaultGroovyMethods.each(DefaultGroovyMethods.java:2040)
	at org.codehaus.groovy.runtime.DefaultGroovyMethods.each(DefaultGroovyMethods.java:2025)
	at org.codehaus.groovy.runtime.DefaultGroovyMethods.each(DefaultGroovyMethods.java:2066)
	at org.codehaus.groovy.runtime.dgm$163.invoke(Unknown Source)
	at org.codehaus.groovy.runtime.callsite.PojoMetaMethodSite$PojoMetaMethodSiteNoUnwrapNoCoerce.invoke(PojoMetaMethodSite.java:274)
	at org.codehaus.groovy.runtime.callsite.PojoMetaMethodSite.call(PojoMetaMethodSite.java:56)
	at org.codehaus.groovy.runtime.callsite.AbstractCallSite.call(AbstractCallSite.java:125)
	at grails.plugins.elasticsearch.ElasticSearchService$_doBulkRequest_closure2.doCall(ElasticSearchService.groovy:282)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.springsource.loaded.ri.ReflectiveInterceptor.jlrMethodInvoke(ReflectiveInterceptor.java:1427)
	at org.codehaus.groovy.reflection.CachedMethod.invoke(CachedMethod.java:93)
	at groovy.lang.MetaMethod.doMethodInvoke(MetaMethod.java:325)
	at org.codehaus.groovy.runtime.metaclass.ClosureMetaClass.invokeMethod(ClosureMetaClass.java:294)
	at groovy.lang.MetaClassImpl.invokeMethod(MetaClassImpl.java:1022)
	at groovy.lang.Closure.call(Closure.java:414)
	at groovy.lang.Closure.call(Closure.java:430)
	at org.codehaus.groovy.runtime.DefaultGroovyMethods.each(DefaultGroovyMethods.java:2040)
	at org.codehaus.groovy.runtime.DefaultGroovyMethods.each(DefaultGroovyMethods.java:2025)
	at org.codehaus.groovy.runtime.DefaultGroovyMethods.each(DefaultGroovyMethods.java:2066)
	at org.codehaus.groovy.runtime.dgm$163.invoke(Unknown Source)
	at org.codehaus.groovy.runtime.callsite.PojoMetaMethodSite$PojoMetaMethodSiteNoUnwrapNoCoerce.invoke(PojoMetaMethodSite.java:274)
	at org.codehaus.groovy.runtime.callsite.PojoMetaMethodSite.call(PojoMetaMethodSite.java:56)
	at org.codehaus.groovy.runtime.callsite.CallSiteArray.defaultCall(CallSiteArray.java:48)
	at org.codehaus.groovy.runtime.callsite.AbstractCallSite.call(AbstractCallSite.java:113)
	at org.codehaus.groovy.runtime.callsite.AbstractCallSite.call(AbstractCallSite.java:125)
	at grails.plugins.elasticsearch.ElasticSearchService.doBulkRequest(ElasticSearchService.groovy:257)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.springsource.loaded.ri.ReflectiveInterceptor.jlrMethodInvoke(ReflectiveInterceptor.java:1427)
	at org.codehaus.groovy.runtime.callsite.PogoMetaMethodSite$PogoCachedMethodSiteNoUnwrapNoCoerce.invoke(PogoMetaMethodSite.java:210)
	at org.codehaus.groovy.runtime.callsite.PogoMetaMethodSite.callCurrent(PogoMetaMethodSite.java:59)
	at org.codehaus.groovy.runtime.callsite.CallSiteArray.defaultCallCurrent(CallSiteArray.java:52)
	at org.codehaus.groovy.runtime.callsite.AbstractCallSite.callCurrent(AbstractCallSite.java:154)
	at org.codehaus.groovy.runtime.callsite.AbstractCallSite.callCurrent(AbstractCallSite.java:174)
	at grails.plugins.elasticsearch.ElasticSearchService.index(ElasticSearchService.groovy:164)
	at grails.plugins.elasticsearch.ElasticSearchBootStrapHelper.bulkIndexOnStartup(ElasticSearchBootStrapHelper.groovy:34)
	at grails.plugins.elasticsearch.ElasticSearchBootStrapHelper$bulkIndexOnStartup.call(Unknown Source)
	at org.codehaus.groovy.runtime.callsite.CallSiteArray.defaultCall(CallSiteArray.java:48)
	at org.codehaus.groovy.runtime.callsite.AbstractCallSite.call(AbstractCallSite.java:113)
	at org.codehaus.groovy.runtime.callsite.AbstractCallSite.call(AbstractCallSite.java:117)
	at org.codehaus.groovy.runtime.callsite.AbstractCallSite.callSafe(AbstractCallSite.java:80)
	at ElasticsearchBootStrap$_closure1.doCall(ElasticsearchBootStrap.groovy:8)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.springsource.loaded.ri.ReflectiveInterceptor.jlrMethodInvoke(ReflectiveInterceptor.java:1427)
	at org.codehaus.groovy.reflection.CachedMethod.invoke(CachedMethod.java:93)
	at groovy.lang.MetaMethod.doMethodInvoke(MetaMethod.java:325)
	at org.codehaus.groovy.runtime.metaclass.ClosureMetaClass.invokeMethod(ClosureMetaClass.java:294)
	at groovy.lang.MetaClassImpl.invokeMethod(MetaClassImpl.java:1022)
	at groovy.lang.MetaClassImpl.invokeMethod(MetaClassImpl.java:1087)
	at groovy.lang.MetaClassImpl.invokeMethod(MetaClassImpl.java:1022)
	at groovy.lang.Closure.call(Closure.java:414)
	at groovy.lang.Closure.call(Closure.java:408)
	at grails.util.Environment.evaluateEnvironmentSpecificBlock(Environment.java:541)
	at grails.util.Environment.executeForEnvironment(Environment.java:534)
	at grails.util.Environment.executeForCurrentEnvironment(Environment.java:510)
	at org.grails.web.servlet.boostrap.DefaultGrailsBootstrapClass.callInit(DefaultGrailsBootstrapClass.java:74)
	at org.grails.web.servlet.context.GrailsConfigUtils.executeGrailsBootstraps(GrailsConfigUtils.java:65)
	at org.grails.plugins.web.servlet.context.BootStrapClassRunner.onStartup(BootStrapClassRunner.groovy:53)
	at grails.boot.config.GrailsApplicationPostProcessor.onApplicationEvent(GrailsApplicationPostProcessor.groovy:261)
	at grails.boot.config.GrailsApplicationPostProcessor.onApplicationEvent(GrailsApplicationPostProcessor.groovy)
	at org.springframework.context.event.SimpleApplicationEventMulticaster.doInvokeListener(SimpleApplicationEventMulticaster.java:172)
	at org.springframework.context.event.SimpleApplicationEventMulticaster.invokeListener(SimpleApplicationEventMulticaster.java:165)
	at org.springframework.context.event.SimpleApplicationEventMulticaster.multicastEvent(SimpleApplicationEventMulticaster.java:139)
	at org.springframework.context.support.AbstractApplicationContext.publishEvent(AbstractApplicationContext.java:393)
	at org.springframework.context.support.AbstractApplicationContext.publishEvent(AbstractApplicationContext.java:347)
	at org.springframework.context.support.AbstractApplicationContext.finishRefresh(AbstractApplicationContext.java:883)
	at org.springframework.boot.context.embedded.EmbeddedWebApplicationContext.finishRefresh(EmbeddedWebApplicationContext.java:144)
	at org.springframework.context.support.AbstractApplicationContext.refresh(AbstractApplicationContext.java:546)
	at org.springframework.boot.context.embedded.EmbeddedWebApplicationContext.refresh(EmbeddedWebApplicationContext.java:122)
	at org.springframework.boot.SpringApplication.refresh(SpringApplication.java:693)
	at org.springframework.boot.SpringApplication.refreshContext(SpringApplication.java:360)
	at org.springframework.boot.SpringApplication.run(SpringApplication.java:303)
	at grails.boot.GrailsApp.run(GrailsApp.groovy:83)
	at grails.boot.GrailsApp.run(GrailsApp.groovy:387)
	at grails.boot.GrailsApp.run(GrailsApp.groovy:374)
	at grails.boot.GrailsApp$run.call(Unknown Source)
	at org.codehaus.groovy.runtime.callsite.CallSiteArray.defaultCall(CallSiteArray.java:48)
	at org.codehaus.groovy.runtime.callsite.AbstractCallSite.call(AbstractCallSite.java:113)
	at org.codehaus.groovy.runtime.callsite.AbstractCallSite.call(AbstractCallSite.java:133)
	at xxx.Application.main(Application.groovy:8)

Here is my build.gradle (as recommended from issue #198):

    def esVersion = '5.4.1'
    ext['elasticsearch.version'] = esVersion

    ext {
        elasticsearchVersion = esVersion
    }
    
    dependencies {
    ...
        compile 'org.grails.plugins:elasticsearch:2.4.0.RC1'
        runtime 'org.elasticsearch.plugin:mapper-attachments:2.4.6'
    ...
    }

Here is my configuration code from application.groovy:

    elasticSearch.datastoreImpl = 'hibernateDatastore'
    elasticSearch.client.mode = 'local'
    elasticSearch.migration.strategy = 'delete'
    elasticSearch.disableAutoIndex = false
    elasticSearch.bulkIndexOnStartup = true

Here is code from my User domain:

    class User implements Serializable {
	    static searchable = {
		    only = ['username', 'firstName', 'lastName', 'description']
	    }
       String username
	   String firstName
	   String lastName
	   String description
       ...
    }

In case the issue is with caching, here is my hibernate configuration in application.groovy:

    hibernate {
	    connection.charSet = 'utf8mb4'
	    connection.characterEncoding='utf8mb4'
	    connection.useUnicode=true
	    cache.use_second_level_cache=true
	    cache.use_query_cache=true
	    cache.provider_class = 'net.sf.ehcache.hibernate.EhCacheProvider'
	    cache.region.factory_class = 'org.hibernate.cache.ehcache.EhCacheRegionFactory' // Hibernate 4+5
	    cache.region_prefix=""
	    singleSession = true 
	    flush.mode = 'manual' 
    }

And my ehcache dependency in build.gradle:

    compile "org.hibernate:hibernate-core:5.1.5.Final"
    compile "org.hibernate:hibernate-ehcache:5.1.5.Final"

### PR DESCRIPTION
**Pull request by dustindclark**
Fixes #214 .

### CODE REVIEW

**Comment by verglor** [comment]
Hi @dustindclark, it seems this PR is not buildable since you have removed publish plugins. Probably you should provide only necessary changes to fix the problem without modifying build process and name the PR accordingly (e.g. Support for Hibernate proxy instances).

**Comment by puneetbehl** [comment]
@verglor I am trying to manually merge this PR and only picking up the required changes.

**Comment by dustindclark** [comment]
I apologize for the inadvertent commit...the build.gradle changes were not supposed to be in the pr.

File: src/main/groovy/grails/plugins/elasticsearch/ElasticSearchContextHolder.groovy [hunk 1]
```diff
@@ -45,7 +46,7 @@ class ElasticSearchContextHolder {
      * @return
      */
     SearchableClassMapping getMappingContext(DomainEntity domainClass) {
```
Comment (puneetbehl) [inline src/main/groovy/grails/plugins/elasticsearch/ElasticSearchContextHolder.groovy]: Though, I have not fully verified but it seems like this change will affect other functionalities too. Are you sure this is required?

**Pull request closed without merging**